### PR TITLE
Add sharding support for dynamic masks in TPU splash attention

### PR DESCRIPTION
**Pull request by Rifur13**
This PR adds support for sharding dynamic masks in TPU splash attention along the head and sequence dimensions. 

### Testing
- Added multi-device tests for both dynamic and static masks

### CODE REVIEW

File: tests/pallas/tpu_splash_attention_kernel_sharded_test.py [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+        in_specs=(
+            kernel_spec,
+            q_spec,
+            kv_spec,
+            kv_spec,
+        ),
+        out_specs=q_spec,
+        check_rep=False,
+    )
+    def f(kernel, q, k, v):
```
[the remark this answers is not in the document]
Comment (Rifur13) [reply tests/pallas/tpu_splash_attention_kernel_sharded_test.py:122]: The API is actually the same for dynamic vs static masks. The `kernel` arg is necessary to be able to shard its MaskInfo with `kernel_spec`. Not that when using static masks you still want to shard `MaskInfo.data_next` and `MaskInfo.mast_next` since they're shapes are static.

[the remark this answers is not in the document]
Comment (Rifur13) [reply tests/pallas/tpu_splash_attention_kernel_sharded_test.py:122]: One idea to make the api simpler is to have a wrapper function that calls `shard_map` and  `manual_sharding_spec` directly. You would only need to pass the PartitionSpecs of the inputs. 

```
out = splash_attention(q, k, v, q_pspec, k_pspec, v_pspec, **kwargs)
```

File: tests/pallas/tpu_splash_attention_kernel_sharded_test.py [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+    devices = np.asarray(jax.devices()[:num_devices]).reshape(
+        head_shards, q_seq_shards
+    )
+
+    mesh = jax.sharding.Mesh(devices, ("heads", "q_seq"))
+    q_spec = PartitionSpec(
+        "heads" if head_shards > 1 else None,
+        "q_seq" if q_seq_shards > 1 else None,
+    )
+    kv_spec = PartitionSpec("heads" if head_shards > 1 else None, None)
```
[the remark this answers is not in the document]
Comment (Rifur13) [reply tests/pallas/tpu_splash_attention_kernel_sharded_test.py:102]: Not exactly, this PR only enables sharding for `MaskInfo.partial_mask_blocks` along the sequence axis **and** the head axis. All other inputs were shardable along these 2 axes before.

File: tests/pallas/tpu_splash_attention_kernel_sharded_test.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (apghml) [reply tests/pallas/tpu_splash_attention_kernel_sharded_test.py:122]: Currently, we wrap a bunch of different kernels including gpu, tpu splash, tpu legacy, flash decoding, etc. in a single unified shard map. So I'm not sure if we would be able to use one specific to splash attention.

Maybe this is actually more of a problem with the shard map / vmap API, but I was trying to figure out whether it is possible to avoid needing two separate branches for this code:

```
    if isinstance(splash_mask, splash_attention_mask.Mask):
        splash_mask = splash_attention_mask.MultiHeadMask(masks=[splash_mask] * num_heads)
        kernel = lambda q, k, v: _make_splash_attention(
            mask=splash_mask,
            ...
        )(q=q, k=k, v=v)
        kernel = jax.vmap(kernel)
        return kernel(q=query, k=key, v=value)
    else: # It is a jax.Array.
        splash_mask = jnp.broadcast_to(splash_mask, (query.shape[0], *splash_mask.shape[1:]))
        kernel = lambda q, k, v, mask: _make_splash_attention(
            mask=mask,
            ...
        )(q=q, k=k, v=v)
        kernel = jax.vmap(kernel)
        return kernel(q=query, k=key, v=value, mask=splash_mask)
```

**Review by sharadmv — APPROVED** [review]

**Review by skye — APPROVED** [review]
